R/revision_analysis.R

```diff
@@ -0,0 +1,209 @@
+      cli_inform("Fraction of all versions that are `NA`:")
+      cli_li(num_percent(total_na, nrow(epi_arch$DT)))
+    }
+    total_num_unrevised <- sum(revision_behavior$n_revisions == 0) # nolint: object_usage_linter
+    cli_inform("No revisions:")
+    cli_li(num_percent(total_num_unrevised, total_num))
+    total_quickly_revised <- sum( # nolint: object_usage_linter
+      revision_behavior$max_lag <=
+        as.difftime(quick_revision, units = "days")
+    )
+    cli_inform("Quick revisions (last revision within {quick_revision}
+{units(quick_revision)} of the `time_value`):")
+    cli_li(num_percent(total_quickly_revised, total_num))
+    total_barely_revised <- sum( # nolint: object_usage_linter
+      revision_behavior$n_revisions <=
+        few_revisions
+    )
+    cli_inform("Few revisions (At most {few_revisions} revisions for that `time_value`):")
+    cli_li(num_percent(total_barely_revised, total_num))
+    cli_inform("")
+    cli_inform("Changes in Value:")
+
+    real_revisions <- revision_behavior %>% filter(n_revisions > 0) # nolint: object_usage_linter
+    n_real_revised <- nrow(real_revisions) # nolint: object_usage_linter
+    rel_change <- sum( # nolint: object_usage_linter
+      real_revisions$max_rel_change <
+        rel_change_threshold,
+      na.rm = TRUE
+    ) + sum(is.na(real_revisions$max_rel_change))
+    cli_inform("Less than {rel_change_threshold} change in relative value (only from the revised subset):")
+    cli_li(num_percent(rel_change, n_real_revised))
+    na_rel_change <- sum(is.na(real_revisions$max_rel_change)) # nolint: object_usage_linter
+    cli_inform("{units(quick_revision)} until over {percent_final_value} percent of the final value:")
+    difftime_summary(revision_behavior[[glue::glue("time_to_pct_final")]])
+    abs_change <- sum( # nolint: object_usage_linter
+      real_revisions$max_change >
+        abs_change_threshold
+    ) # nolint: object_usage_linter
+    cli_inform("Change by more than {abs_change_threshold} in actual value (when revised):")
+    cli_li(num_percent(abs_change, n_real_revised))
+  }
+  return(revision_behavior)
+}
+
+#' @keywords internal
+time_to_x_percent <- function(lags, values, percent = .9) {
+  final_value <- values[which.max(lags)]
+  close_enough <- abs(values - final_value) < (1 - percent) * final_value
+  # we want to ignore any stretches where it's close, but goes farther away later
+  return(get_last_run(close_enough, lags))
+}
+
+#' return the first value in values_from from the last string of trues in bool_vec
+#' @description
+#' the point of this operation is to get the value in values_from which occurs
+#'   at the same index as the start of the last run of true values in bool_vec.
+#'   for example, in c(1,1,0,1,1), we want the 4th entry, since there's a 0
+#'   breaking the run
+#' @keywords internal
+get_last_run <- function(bool_vec, values_from) {
+  runs <- rle(bool_vec)
```

Comment:
idea: if you find this being slow, helper functions from the latency vignette WIP may be helpful.  (based on `rev` + `match`)

Reply:
this is basically a fancy wrapper on rle, which is a base R function. It's not slow enough to be bothersome.